common/flatpak-run.c

```diff
@@ -3165,6 +3169,91 @@ regenerate_ld_cache (GPtrArray    *base_argv_array,
+static gboolean
+check_parental_controls (const char     *app_ref,
+                         FlatpakDeploy  *deploy,
+                         GCancellable   *cancellable,
+                         GError        **error)
+{
+#ifdef HAVE_LIBMALCONTENT
+  g_auto(GStrv) app_ref_parts = NULL;
+  g_autoptr(MctManager) manager = NULL;
+  g_autoptr(MctAppFilter) app_filter = NULL;
+  g_autoptr(GAsyncResult) app_filter_result = NULL;
+  g_autoptr(GDBusConnection) system_bus = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GDesktopAppInfo) app_info = NULL;
+  gboolean allowed = FALSE;
+
+  app_ref_parts = flatpak_decompose_ref (app_ref, error);
+  if (app_ref_parts == NULL)
+    return FALSE;
+
+  system_bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+  if (system_bus == NULL)
+    return FALSE;
+
+  manager = mct_manager_new (system_bus);
+  app_filter = mct_manager_get_app_filter (manager, getuid (),
+                                           MCT_GET_APP_FILTER_FLAGS_INTERACTIVE,
+                                           cancellable, &local_error);
+  if (g_error_matches (local_error, MCT_APP_FILTER_ERROR, MCT_APP_FILTER_ERROR_DISABLED))
+    {
+      g_debug ("Skipping parental controls check for %s since parental "
+               "controls are disabled globally", app_ref);
+      return TRUE;
+    }
+  else if (local_error != NULL)
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+    }
+
+  /* Look up the app’s main .desktop file. */
+  if (deploy != NULL)
+    {
+      g_autoptr(GFile) deploy_dir = NULL;
+      const char *deploy_path;
+      g_autofree char *desktop_file_name = NULL;
+      g_autofree char *desktop_file_path = NULL;
+
+      deploy_dir = flatpak_deploy_get_dir (deploy);
+      deploy_path = flatpak_file_get_path_cached (deploy_dir);
+
+      desktop_file_name = g_strconcat (app_ref_parts[1], ".desktop", NULL);
+      desktop_file_path = g_build_path (G_DIR_SEPARATOR_S,
+                                        deploy_path,
+                                        "export",
+                                        "share",
+                                        "applications",
+                                        desktop_file_name,
+                                        NULL);
+      app_info = g_desktop_app_info_new_from_filename (desktop_file_path);
+    }
+
+  /* Always filter by app ID. Additionally, filter by app info (which runs
+   * multiple checks, including whether the app ID, executable path and
+   * content types are allowed) if available. If the flatpak contains
+   * multiple .desktop files, we use the main one. The app ID check is
+   * always done, as the binary executed by `flatpak run` isn’t necessarily
+   * extracted from a .desktop file. */
+  allowed = mct_app_filter_is_flatpak_ref_allowed (app_filter, app_ref);
+  if (app_info != NULL)
+    allowed = allowed && mct_app_filter_is_appinfo_allowed (app_filter,
```

Comment:
You may be able to save some CPU cycles here if you skip the app_info check (and even building the app_info above) if we already know the ref is not allowed.

Reply:
It makes the code a bit messier, but probably worth it to avoid reading in the desktop file.

Reply:
Why is there separate filtering on the app id and the appinfo in the first place?